go/vt/mysqlctl/builtinbackupengine.go

```diff
@@ -604,7 +604,15 @@ func (be *BuiltinBackupEngine) backupFiles(
 	wg := sync.WaitGroup{}
 
 	ctxCancel, cancel := context.WithCancel(ctx)
-	defer cancel()
+	defer func() {
+		// We may still have operations in flight that require a valid context, such as adding files to S3.
+		// Unless we encountered an error, we should not cancel the context, this is taken care of later
+		// in the process. If we encountered an error however, we can safely cancel the context as we should
+		// no longer work on anything and exit fast.
+		if finalErr != nil {
+			cancel()
+		}
+	}()
```

Comment:
I'm not sure I follow:
- It looks as if the function can exit and still leave the context active, in which case, what is cancelling it?
- If we still have operations in flight, why would we exit the function? Should we not wait until everything is complete?
Something feels off here, as an anti-pattern.

Reply:
> what is cancelling it?

The caller of ExecuteBackup will eventually cancel it, whether it is through a gRPC call or through vtbackup, the context always gets canceled.

Reply:
Thank you for the clarity! ❤️